tachyon/zk/lookup/log_derivative_halo2/prover_impl.h

```diff
@@ -192,34 +192,42 @@ template <typename PCS>
 BlindedPolynomial<Poly, Evals> Prover<Poly, Evals>::CreateGrandSumPoly(
     ProverBase<PCS>* prover, const Evals& m_values,
     const std::vector<Evals>& compressed_inputs, const Evals& compressed_table,
-    const F& beta) {
+    const F& beta, GrandSumPolysTempStorage<F>& storage) {
   size_t n = prover->pcs().N();
+  RowIndex usable_rows = prover->GetUsableRows();
 
   // Σ 1/(φᵢ(X))
-  std::vector<F> inputs_log_derivatives(n, F::Zero());
-  std::vector<F> input_log_derivatives(
-      compressed_inputs[0].evaluations().size());
-  for (const Evals& compressed_input : compressed_inputs) {
-    ComputeLogDerivatives(compressed_input, beta, input_log_derivatives);
-
-    OPENMP_PARALLEL_FOR(size_t i = 0; i < n; ++i) {
-      inputs_log_derivatives[i] += input_log_derivatives[i];
+  // NOTE(chokobole): This is intentional to assign
+  // |storage.table_log_derivatives| to |input_log_derivatives| since
+  // |input_log_derivatives| is not used after this loop.
+  std::vector<F>& input_log_derivatives = storage.table_log_derivatives;
+  for (size_t i = 0; i < compressed_inputs.size(); ++i) {
+    ComputeLogDerivatives(compressed_inputs[i], beta, input_log_derivatives);
+
+    if (i == 0) {
+      OPENMP_PARALLEL_FOR(size_t j = 0; j < usable_rows; ++j) {
+        storage.inputs_log_derivatives[j] = input_log_derivatives[j];
+      }
+    } else {
+      OPENMP_PARALLEL_FOR(size_t j = 0; j < usable_rows; ++j) {
+        storage.inputs_log_derivatives[j] += input_log_derivatives[j];
+      }
     }
   }
 
   // 1 / τ(X)
-  std::vector<F> table_log_derivatives(compressed_table.evaluations().size());
-  ComputeLogDerivatives(compressed_table, beta, table_log_derivatives);
+  ComputeLogDerivatives(compressed_table, beta, storage.table_log_derivatives);
 
   std::vector<F> grand_sum(n);
   grand_sum[0] = F::Zero();
 
   // (Σ 1/φᵢ(X)) - m(X) / τ(X)
-  RowIndex usable_rows = prover->GetUsableRows();
-  std::vector<F> log_derivatives_diff(usable_rows);
+  // NOTE(chokobole): This is intentional to assign
+  // |storage.inputs_log_derivatives| to |log_derivatives_diff| since
+  // |storage.inputs_log_derivatives| is not used after this loop.
```

Comment:
```suggestion
  // NOTE(chokobole): To save memory, |log_derivatives_diff| overwrites
  // |storage.inputs_log_derivatives| since the current values of
  // |storage.inputs_log_derivatives| are not needed anymore.